### PR TITLE
Implemented markup bundle

### PR DESCRIPTION
**Pull request by wachterjohannes**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | none
| Related issues/PRs | https://github.com/sulu/sulu-standard/pull/681
| License | MIT
| Documentation PR | https://github.com/sulu/sulu-docs/pull/218 

#### What's in this PR?

This PR introduces a response-listener which replaces "special" html tags with others. It is extensible via services which gets called with the attributes of the tag to replace.

Inspired by http://www.jochenfroehlich.com/tutorial/typo3-link.html

#### Why?

It should be possible to use html-tags to specify a internal link.

#### Example Usage

~~~html
<html>
    <body>
        <sulu:link href="123-123-123" title="test-title" />
    </body>
</html>
~~~

~~~php
class LinkTag implements TagInterface
{
    /**
     * Returns new tag with given attributes.
     *
     * @param array $attributes attributes array of each tag occurrence.
     *
     * @return array Tag array to replace all occurrences.
     */
    public function parseAll($attributesByTag) {
        $result = [];
        foreach($attributesByTag as $tag => $attributes) {
            $url = ; // load url via uuid from document-manager
            $pageTitle = ...; // load page-title via uuid from document-manager

            $result[$tag] = sprintf('<a href="%s" title="%s">%s</a>', $url, $$pageTitle, $attributes['content']);
        }

        return $result;
    }
}
~~~

Results into:

~~~html
<html>
    <body>
        <a href="/test" title="test-title">Page Title</a>
    </body>
</html>
~~~

#### To Do

- [x] Create a documentation PR
- [x] bundle config and compiler pass

### CODE REVIEW

File: src/Sulu/Bundle/MarkupBundle/Tag/TagInterface.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /**
+     * Returns new tag with given attributes.
+     *
+     * @param string $tag
+     * @param array $attributes
+     *
+     * @return mixed
+     */
+    public function parse($tag, $attributes);
```
Comment (chirimoya) [inline src/Sulu/Bundle/MarkupBundle/Tag/TagInterface.php]: `parseAll`

**Comment by alexander-schranz** [comment]
Just asking why not use the ESI tag specification for this?

https://en.wikipedia.org/wiki/Edge_Side_Includes#How_ESI_is_implemented

Ok its not as beautiful with a query string but its specified tag.

**Comment by wachterjohannes** [comment]
@alexander-schranz we dont want to include another page. we want to "preprocess" the generated html output and for example add links to other pages or generate download links.

**Comment by alexander-schranz** [comment]
@wachterjohannes i see that as page fragments and that is what ESI was made for.

**Comment by wachterjohannes** [comment]
would you only use the syntax or the "system" behind. because only "stealing" the syntax is not a good idea. and i am not sure how to use the system behind. /cc @chirimoya

**Comment by alexander-schranz** [comment]
Not only the syntax also the system because ESI should work without problems in symfony (when enabled?) as they use it in `render_esi`. So we would define a `controller` for this where you can give a required `type` in your example `link` and the other params `href`. The controller will automaticlly be called by symfony. And you call then your `LinkTag` class.

**Comment by chirimoya** [comment]
@alexander-schranz esi would definitely be a big overhead compared to this solution. And its not really a fragment, it's more like a markup extension.

File: src/Sulu/Bundle/MarkupBundle/Listener/MarkupListener.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * Parses content of response and set the replaced html as new content.
+     *
+     * @param FilterResponseEvent $event
+     */
+    public function replaceMarkup(FilterResponseEvent $event)
+    {
+        $response = $event->getResponse();
+        if (0 !== strpos($response->headers->get('Content-Type'), 'text/html')
+            && !$event->getRequest()->get('preview')
```
Comment (danrot) [inline src/Sulu/Bundle/MarkupBundle/Listener/MarkupListener.php]: What kind of influence does the preview have here?

File: src/Sulu/Bundle/MarkupBundle/Markup/MarkupParser.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ */
+
+namespace Sulu\Bundle\MarkupBundle\Markup;
+
+use Sulu\Bundle\MarkupBundle\Tag\TagRegistryInterface;
+
+/**
+ * Parses html content and replaces special tags.
+ */
+class MarkupParser implements MarkupParserInterface
```
Comment (danrot) [inline src/Sulu/Bundle/MarkupBundle/Markup/MarkupParser.php]: Should we maybe call that `HtmlMarkupParser`, in case we also add something similar for e.g. XML or RSS in the future?

File: src/Sulu/Bundle/MarkupBundle/Markup/MarkupParser.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+{
+    const ATTRIBUTE_REGEX = '/(?<name>\b\w+\b)\s*=\s*"(?<value>[^"]*)"/';
+    const CONTENT_REGEX = '/(?:>(?<content>.*)<)/';
+    const TAG_REGEX = '/(?<tag><%s:(?<name>[a-z]*)[^\/>]*(?:\/>|>.*<\/%s:[^\/>]*>))/';
+
+    /**
+     * @var TagRegistryInterface
+     */
+    private $tagRegistry;
+    /**
```
Comment (danrot) [inline src/Sulu/Bundle/MarkupBundle/Markup/MarkupParser.php]: New line before this line.

File: src/Sulu/Bundle/MarkupBundle/Markup/MarkupParser.php [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        $this->tagRegistry = $tagRegistry;
+        $this->namespace = $namespace;
+    }
+
+    /**
+     * @param string $html
+     *
+     * @return string
+     */
+    public function parse($html)
```
Comment (danrot) [inline src/Sulu/Bundle/MarkupBundle/Markup/MarkupParser.php]: If we apply my previous suggestion `$html` should probably be renamed to `$html`.

File: src/Sulu/Component/Content/Repository/ContentRepository.php [hunk 6]
```diff
@@ -186,7 +186,7 @@ public function findParentsWithSiblingsByUuid(
             ->orderBy($this->qomFactory->propertyValue('node', 'jcr:path'))
             ->where($this->qomFactory->childNode('node', $path));
 
-        while ($path !== $contentPath) {
+        while (PathHelper::getPathDepth($path) > PathHelper::getPathDepth($contentPath)) {
```
Comment (danrot) [inline src/Sulu/Component/Content/Repository/ContentRepository.php]: Somehow connected to the `MarkupBundle`? Or what is this doing?

File: src/Sulu/Bundle/MarkupBundle/Listener/MarkupListener.php (hunk 2 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/MarkupBundle/Listener/MarkupListener.php]: for preview the mime-type will not be set in websocket. i will check this on tuesday

File: src/Sulu/Component/Content/Repository/ContentRepository.php (hunk 6 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Content/Repository/ContentRepository.php]: yes and no (: this should be in the internal-link PR because there you can mix the webspaces and then the check has to be changed like this. i will move it to the internal-link PR.

File: src/Sulu/Bundle/MarkupBundle/Markup/MarkupParser.php (hunk 5 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/MarkupBundle/Markup/MarkupParser.php]: fixed

File: src/Sulu/Bundle/MarkupBundle/Markup/MarkupParser.php (hunk 4 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/MarkupBundle/Markup/MarkupParser.php]: fixed

**Comment by wachterjohannes** [comment]
@danrot when the tests are green this could be merged

File: src/Sulu/Bundle/MarkupBundle/Markup/MarkupParserInterface.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\MarkupBundle\Markup;
+
+/**
+ * Parses html content and replaces special tags.
```
Comment (danrot) [inline src/Sulu/Bundle/MarkupBundle/Markup/MarkupParserInterface.php]: Remove html

File: src/Sulu/Bundle/MarkupBundle/Markup/MarkupParserInterface.php [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ */
+interface MarkupParserInterface
+{
+    /**
+     * Parses html-document and returns completed html.
```
Comment (danrot) [inline src/Sulu/Bundle/MarkupBundle/Markup/MarkupParserInterface.php]: rephrase to avoid using html

File: src/Sulu/Bundle/MarkupBundle/Tag/TagRegistry.php [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        $this->tags = $tags;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTag($name, $type)
+    {
+        if (!array_key_exists($type, $this->tags) || !array_key_exists($name, $this->tags[$type])) {
+            throw new TagNotFoundException($name);
```
Comment (danrot) [inline src/Sulu/Bundle/MarkupBundle/Tag/TagRegistry.php]: Shouldn't this be called `TagTypeNotFoundException`?

Comment (danrot) [inline src/Sulu/Bundle/MarkupBundle/Tag/TagRegistry.php]: Sry, have overseen the second condition... But then you should also pass the type here, and adapt the Exception accordingly.

File: src/Sulu/Bundle/MarkupBundle/Listener/MarkupListener.php [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    private $mimeTypeFormatMap;
+
+    /**
+     * @param MarkupParserInterface[] $markupParser
+     * @param array $mimeTypeFormatMap
+     */
+    public function __construct(
+        array $markupParser,
+        array $mimeTypeFormatMap = ['text/html' => 'html', 'application/xml' => 'xml', 'text/xml' => 'xml']
```
Comment (danrot) [inline src/Sulu/Bundle/MarkupBundle/Listener/MarkupListener.php]: I would not pass this array here, because it is not extendable. For this reason I would place the defaults in the DI configuration.

**Comment by danrot** [comment]
Can you integrate the `sulu:link` from the example in this PR, or put that at least in a very own PR, which is based on this one? (personally I would prefer first variant, then we should maybe also rename the PR).

I'd like to review this in separate.

**Comment by wachterjohannes** [comment]
@danrot when the tests are green this PR contains the fixes you mentioned and the link-tag

File: src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php [hunk 11]
```diff
@@ -155,7 +155,7 @@ public function render($object, $id, $webspaceKey, $locale, $partial = false)
         $this->eventDispatcher->dispatch(Events::PRE_RENDER, new PreRenderEvent($attributes));
 
         try {
-            $response = $this->httpKernel->handle($request, HttpKernelInterface::SUB_REQUEST, false);
+            $response = $this->httpKernel->handle($request, HttpKernelInterface::MASTER_REQUEST, false);
```
Comment (danrot) [inline src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php]: Why was that necessary?

File: src/Sulu/Bundle/MarkupBundle/Resources/config/services.xml [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+            <tag name="sulu_markup.parser" type="html"/>
+        </service>
+
+        <service id="sulu_markup.response_listener" class="Sulu\Bundle\MarkupBundle\Listener\MarkupListener">
+            <argument type="collection"/>
+            <argument type="collection">
+                <argument type="string" key="text/html">html</argument>
+                <argument type="string" key="text/xml">xml</argument>
+                <argument type="string" key="application/xml">xml</argument>
```
Comment (danrot) [inline src/Sulu/Bundle/MarkupBundle/Resources/config/services.xml]: I would move that somehow to the `Configuration` and `Extension` class. Or can we maybe even use [Symfony builtins](http://symfony.com/doc/current/cookbook/request/mime_type.html) instead?

**Comment by danrot** [comment]
Two comments are still open, and the tests are failing.

File: src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php (hunk 11 above)
Comment (danrot) [inline src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php]: Please answer that question, so that I can understand 😄

File: src/Sulu/Bundle/MarkupBundle/Markup/MarkupParserInterface.php (hunk 7 above)
Comment (danrot) [inline src/Sulu/Bundle/MarkupBundle/Markup/MarkupParserInterface.php]: Still todo.

File: src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php (hunk 11 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php]: to be sure that all listener are called the request has to be rendered as master-request. this makes currently problems with the websocket. i have to check this tomorrow.

**Comment by wachterjohannes** [comment]
@danrot as i said the preview in websocket mode makes currently problems i will take a look to this tomorrow.

**Comment by wachterjohannes** [comment]
i have to disable the profiler (could use the matcher mechanism) http://symfony.com/doc/current/cookbook/profiler/matchers.html#creating-a-custom-matcher

File: src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php (hunk 11 above)
Comment (danrot) [inline src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php]: If you'd like to make sure it works for both kind of request you would also have to test both variants, not change the existing one 😉

File: src/Sulu/Bundle/MarkupBundle/Markup/MarkupParserInterface.php (hunk 7 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/MarkupBundle/Markup/MarkupParserInterface.php]: sorry i have overseen this html. should be fixed now
